pkg/flow/internal/controller/node_component.go

```diff
@@ -257,6 +258,13 @@ func (cn *ComponentNode) Evaluate(scope *vm.Scope) error {
 		msg := fmt.Sprintf("component evaluation failed: %s", err)
 		cn.setEvalHealth(component.HealthTypeUnhealthy, msg)
 	}
+	// Check for failure on initial loading, if not then we need to remove any modules it the component may habe created.
+	// In the case where the component is not a module loader this is a noop.
+	cn.moduleFailureCheck.Do(func() {
+		if err != nil {
+			cn.moduleController.ClearModuleIDs()
+		}
+	})
```

Comment:
There's still a problem here: what if a component fails to be constructed more than once? 

Rather than using a sync.Once, this check could be moved down to line 289 (i.e., if `cn.reg.Build` fails) so you only clear module IDs for components which can't be built, no matter how many times the construction fails.